### PR TITLE
[Kernel] Add SM120 per-tensor FP8 GEMM for small-M (#33632)

### LINKED ISSUE #33632: [Feature] Optimize Per-Tensor FP8 GEMM on SM120
## Checklist

- [x] If this is not a feature request but a general question, please start a discussion at https://github.com/sgl-project/sglang/discussions. Otherwise, it will be closed.
- [x] Please use English. Otherwise, it will be closed.

## Motivation

Per-tensor FP8 GEMM might not be optimal on SM120.

The shapes below are from `nvidia/Qwen3.6-27B-NVFP4` at TP=1. This checkpoint is mixed precision: the attention groups are per-tensor FP8, and the MLP is NVFP4 using BF16 activations. Only the attention groups are listed here.

**`nvidia/Qwen3.6-27B-NVFP4` — per-tensor FP8 GEMM shapes (TP=1)**

Hidden size 5120, 64 layers (48 linear attention + 16 full attention).

| GEMM | (N, K) | layers |
|---|---|---|
| `linear_attn.in_proj_qkvz` | 16384, 5120 | ×48 |
| `linear_attn.out_proj` | 5120, 6144 | ×48 |
| `self_attn.qkv_proj` | 14336, 5120 | ×16 |
| `self_attn.o_proj` | 5120, 6144 | ×16 |

**`in_proj_qkvz` and `out_proj` are the most important shapes to improve.** They run in 48 of the 64 layers, so they are most of the per-tensor FP8 GEMM calls in the model.

The checkpoint ships a single weight scale and a single activation scale per tensor.

For M: under MTP, M will be a multiple of 4, and we can consider M < 64 (so at most 16 running requests under MTP).

## Which backend to change

Currently we are using cuBLAS for per-tensor FP8 on SM120.

Upgrading cuBLAS to the very latest version will use kernels with TMA, but the gain is not significant for these shapes.

## Measuring perf

Only use `flashinfer.testing.utils.bench_gpu_time_with_cupti` with CUDA graph on to measure perf.

## Related resources

_(none yet)_

### PR DESCRIPTION
**Pull request by evanderfff123-boop**
## Motivation
Fixes [#33632](https://github.com/sgl-project/sglang/issues/33632).

On SM120, ModelOpt per-tensor FP8 linears currently go through `flashinfer.bmm_fp8(..., backend="cublas")`. For the attention-group shapes from `nvidia/Qwen3.6-27B-NVFP4` at TP=1 and small M (MTP, `M < 64`), a dedicated in-tree CUTLASS dense GEMM is faster than cuBLAS on a useful subset of those shapes. This PR adds that kernel and routes only the profitable shapes to it.

## Modifications
- Add an SM120 JIT per-tensor FP8 scaled GEMM (`fp8_pertensor_scaled_mm`) under `sglang.kernels` with two Pingpong tiles (`64x128x64` for `M < 24`, `64x64x128` otherwise) and a fused dual-scale epilogue. 
- Route profitable shapes through the new kernel in `apply_fp8_linear_bmm_flashinfer` / `flashinfer_bmm_fp8`; otherwise keep
cuBLAS.
- `is_profitable`:
    - wide N (`N > 8192`): `M >= 16`
    - narrow N: `M >= 24` (do not route `M < 24`; GEMM-only CUPTI sweep shows cuBLAS ahead by ~9–11% there)
- Add correctness test, production dispatch test, and a registered benchmark using `bench_gpu_time_with_cupti` with CUDA graph on and cold L2, matching the issue measurement requirement.

## Accuracy Tests

```bash
python -m pytest test/registered/kernels/ops/gemm/test_fp8_pertensor_gemm.py -q
......                                                                          [100%]
================================== warnings summary ===================================
../../../../usr/local/lib/python3.12/dist-packages/_pytest/config/__init__.py:1464
  /usr/local/lib/python3.12/dist-packages/_pytest/config/__init__.py:1464: PytestConfigWarning: Unknown config option: asyncio_mode
  
    self._warn_or_fail_if_strict(f"Unknown config option: {key}\n")

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
6 passed, 1 warning in 6.51s

python -m pytest test/registered/unit/layers/quantization/test_fp8_pertensor_dispatch.py -q
.                                                                               [100%]
================================== warnings summary ===================================
../../../../usr/local/lib/python3.12/dist-packages/_pytest/config/__init__.py:1464
  /usr/local/lib/python3.12/dist-packages/_pytest/config/__init__.py:1464: PytestConfigWarning: Unknown config option: asyncio_mode
  
    self._warn_or_fail_if_strict(f"Unknown config option: {key}\n")

../../../../usr/local/lib/python3.12/dist-packages/torch/jit/_script.py:365: 14 warnings
  /usr/local/lib/python3.12/dist-packages/torch/jit/_script.py:365: DeprecationWarning: `torch.jit.script_method` is deprecated. Please switch to `torch.compile` or `torch.export`.
    warnings.warn(

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
1 passed, 15 warnings in 7.77s
```

## Speed Tests and Profiling
Method (per #33632): `bench_gpu_time_with_cupti` with CUDA graph on + cold L2, GEMM only.
Hardware: RTX PRO 5000 72GB Blackwell (SM120).
The registered bench sweeps all issue shapes; production only uses shapes where `is_profitable` is true (wide N: `M >= 16`; narrow N: `M >= 24`). Rows below are the **routed** subset. Ratio = ours / cuBLAS (`< 1` means ours faster).

### Routed results
| N | K | M | ours (µs) | cuBLAS (µs) | ratio |
|---|---|---|---|---|---|
| 16384 | 5120 | 16 | 93.57 | 119.62 | **0.78** |
| 16384 | 5120 | 24 | 83.20 | 120.67 | **0.69** |
| 16384 | 5120 | 32 | 83.81 | 94.34 | **0.89** |
| 16384 | 5120 | 48 | 84.67 | 85.31 | 0.99 |
| 16384 | 5120 | 60 | 85.38 | 85.15 | 1.00 |
| 5120 | 6144 | 24 | 33.70 | 54.91 | **0.61** |
| 5120 | 6144 | 32 | 33.34 | 50.18 | **0.66** |
| 5120 | 6144 | 48 | 34.69 | 58.45 | **0.59** |
| 5120 | 6144 | 60 | 34.21 | 57.15 | **0.60** |
| 14336 | 5120 | 16 | 85.47 | 97.60 | **0.88** |
| 14336 | 5120 | 24 | 78.11 | 98.34 | **0.79** |
| 14336 | 5120 | 32 | 78.59 | 83.14 | **0.95** |
| 14336 | 5120 | 48 | 78.85 | 88.86 | **0.89** |

Unrouted small-M points (e.g. narrow-N M=4/8/16) are slower than cuBLAS by ~9–11% and are intentionally left on the cuBLAS path.
## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [x] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [x] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [x] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

cc @b8zhong 

---
### CI States

Latest PR Test (Base): :x: [Run #34763737183](https://github.com/sgl-project/sglang/actions/runs/34763737183)
Latest PR Test (Extra): :x: [Run #34763737073](https://github.com/sgl-project/sglang/actions/runs/34763737073)
Latest PR Test (AMD ROCm 10): :x: [Run #34763737274](https://github.com/sgl-project/sglang/actions/runs/34763737274)

### CODE REVIEW

**Comment by evanderfff123-boop** [comment]
Hi @b8zhong — gentle ping on this PR for #33632.
Summary of what's in:
- In-tree SM120 per-tensor FP8 GEMM (CUTLASS Pingpong, fused dual-scale epilogue)
- Production routing via `is_profitable` (wide N: M>=16; narrow N: M>=24)
- Correctness + dispatch tests, and a CUPTI + CUDA-graph + cold-L2 bench matching the issue methodology
- Routed shapes are faster or tied vs cuBLAS; unprofitable small-M stays on cuBLAS
Could you take a look when you have a moment, and if needed help with `/tag-and-rerun-ci`?
Happy to address any review comments. Thanks!

**Comment by b8zhong** [comment]
Hi, do you know the E2E speedup for 27B NVFP4?

**Comment by evanderfff123-boop** [comment]
> Hi, do you know the E2E speedup for 27B NVFP4?

Hi @b8zhong, thanks for asking about E2E — it was the right question.

I ran it on Qwen3.6-27B-NVFP4 (TP=1, MTP, decode-heavy), with an A/B that toggles only this kernel's routing. The end-to-end speedup was negligible — within run-to-run noise.

Looking into why, the issues were mostly mine:

I optimized to the CUPTI cold-L2 microbenchmark as the goal rather than as a proxy, and never checked that it predicts real serving. Measured inside an actual forward, the margin over cuBLAS was much smaller.
My is_profitable thresholds gate out the smallest M, which under MTP at low concurrency is the common case — so the fast path was barely active in the regime this issue targets.
The separate low-M tile turns out to be slower than the other tile at essentially every small M, so that branch was hurting rather than helping.
I never profiled where this op sits in a decode step before tuning it, so I had no basis for how much gain was reachable.
This is my first contribution here, so thanks for the patience. I'm reworking the kernel — the small-M path in particular — rather than defending the current version, and will report back with both kernel-level and E2E numbers once confirmed.

**Comment by b8zhong** [comment]
Hi, thanks for replying. (By the way it's usually easier to summarize the results in your own words, which makes it easier for maintainers to understand). Thus, I think we don't need this PR. What do you think?

**Comment by evanderfff123-boop** [comment]
> Hi, thanks for replying. (By the way it's usually easier to summarize the results in your own words, which makes it easier for maintainers to understand). Thus, I think we don't need this PR. What do you think?

Hi, thanks for pointing this out. I admit that I used AI to summarize the issue back then, which made the content obscure and caused inconvenience to the maintainers. I’m very sorry.

To summarize in my own words: I ran end-to-end tests on NvFP4 27B using both my optimized kernel and the original unoptimized kernel. The end-to-end test results show that the speedup barely changed.

The reason is that I observed a small improvement at the operator level, but this minor gain got diluted in the end-to-end scenario. I have also identified the root cause: the kernel optimization is incomplete, and many small-M shapes are not covered. I will work on another optimized version later.

I’m really sorry for the trouble. If I achieve obvious end-to-end performance gains, I will document the entire process, optimizations and benchmark results fully in my own words. Could you please give me another chance? Or at least wait for my next set of results before deciding whether to close this PR, if that’s okay.

**Comment by b8zhong** [comment]
Sure, that's fine. Thanks a lot!

**Comment by evanderfff123-boop** [comment]
> Sure, that's fine. Thanks a lot!

Thanks！

**Comment by evanderfff123-boop** [comment]
Hi, I have a few questions. I have not done end-to-end kernel tuning on this framework before, so I am a bit confused:

First, the kernel I am optimizing is currently written entirely by me in CUTLASS, or I may later switch it to another higher-performance implementation. Can it be written in Triton?

The issue is about optimizing four shapes for Qwen3.6 27B NVFP4 with TP = 1 and M < 64. When routing, can my new kernel be routed specifically to these four shapes? I do not know whether my kernel would be slower than the original one on other shapes. Or is a more general kernel needed? If it is the former, can I do something like: only use my kernel when M < 64 and shape = linear_attn.in_proj_qkvz, etc.?

If a more general kernel is needed, can I add a switch in environ.py to turn my new kernel on and off, so that I can easily disable it later if there is a performance regression on other shapes?

**Comment by evanderfff123-boop** [comment]
Hi, @b8zhong, thank you very much for reading my comment. This optimization mainly uses SwapAB, and picks between two warp-specialized scheduling modes — Cooperative (Coop) and Pingpong (Ping) — depending on the size of M; I've also done end-to-end testing. On the routing side, our kernel is currently invoked only when M ≤ 64, and I've added a flag in environ.py to control whether our kernel is used. I also removed the fairly self-evident dispatch test. Below are the results I measured on a single GPU, including the accuracy test, the kernel benchmark, and the E2E test. If there's anything that needs improvement or is lacking, I'd greatly appreciate your advice and guidance. Thanks again! By the way, I noticed my previous comment hasn't been replied to yet — if you have some time, I'd appreciate a reply as well. Thank you!

### How it was tested
Hardware:  **NVIDIA RTX PRO 5000 Blackwell (SM120, 72 GB)**

1. **Accuracy** — `test/registered/kernels/ops/gemm/test_fp8_pertensor_gemm.py`. Compares the kernel against an fp32-dequant reference (rtol 0.02 / atol 1) over M ∈ {4, 16, 32, 33, 48, 64} × (N,K) ∈ {(16384,5120), (5120,6144), (14336,5120)}.
2. **Kernel benchmark** — `test/registered/kernels/benchmark/gemm/bench_fp8_pertensor_gemm.py`. Median µs, cold-L2 + CUDA graph. **Ours** = our SM120 per-tensor kernel; **cuBLAS** = FlashInfer `bmm_fp8` (`backend="cublas"`).
3. **End-to-end** — `sglang.launch_server` on `nvidia/Qwen3.6-27B-NVFP4`, TP1, NEXTN spec (`num_draft_tokens=4`), `sglang.bench_serving` with `random-ids` 256in/512out, concurrency 1/4/8/16. Verify batch M ≈ concurrency × 4. 

### Results

**Accuracy:** 18/18 passed.

**Kernel benchmark** (median µs; `×` = cuBLAS / Ours):

| M | N=16384,K=5120 (Ours/cuBLAS/×) | N=5120,K=6144 (Ours/cuBLAS/×) | N=14336,K=5120 (Ours/cuBLAS/×) |
|---:|---|---|---|
| 4  | 79.3 / 91.8 / **1.16×** | 32.1 / 34.1 / **1.06×** | 74.3 / 79.6 / **1.07×** |
| 8  | 79.5 / 93.4 / **1.18×** | 31.6 / 34.1 / **1.08×** | 74.4 / 80.4 / **1.08×** |
| 16 | 79.5 / 118.4 / **1.49×** | 31.6 / 36.1 / **1.14×** | 74.4 / 96.3 / **1.29×** |
| 24 | 78.7 / 119.5 / **1.52×** | 31.7 / 53.9 / **1.70×** | 74.9 / 97.4 / **1.30×** |
| 32 | 79.3 / 93.5 / **1.18×** | 31.8 / 50.7 / **1.59×** | 74.7 / 82.0 / **1.10×** |
| 48 | 86.8 / 84.6 / 0.97× | 32.0 / 57.6 / **1.80×** | 79.1 / 87.1 / **1.10×** |
| 60 | 87.8 / 85.1 / 0.97× | 31.5 / 56.7 / **1.80×** | 80.0 / 88.2 / **1.10×** |

Our kernel leads across the M ≤ 64 decode range, up to about 1.80×. The only weak spot is large N with the larger M values: at N=16384 and M=48/60, it is about 3% slower than cuBLAS.

**End-to-end** (throughput ×, TPOT, e2e latency; Ours vs cuBLAS):

| concurrency (≈M) | out tok/s (Ours) | out tok/s (cuBLAS) | throughput × | median TPOT (Ours/cuBLAS ms) | median e2e (Ours/cuBLAS ms) |
|---:|---:|---:|:--:|---|---|
| 1  (≈4)  | 120.5 | 115.2 | 1.05× | 7.85 / 8.51  | 4103 / 4468 |
| 4  (≈16) | 405.2 | 392.8 | 1.03× | 9.07 / 9.33  | 4787 / 4901 |
| 8  (≈32) | 682.0 | 668.9 | 1.02× | 10.88 / 10.99 | 5720 / 5805 |
| 16 (≈64) | 958.2 | 938.9 | 1.02× | 15.54 / 15.56 | 8154 / 8331 |

End-to-end improvement is 2–5%, more pronounced at lower concurrency (at C=1, TPOT drops ~8%, tapering toward ~0 at high concurrency). The direction is consistent with the kernel benchmark (Ours is never slower than cuBLAS), but the overall magnitude is far smaller than the kernel's per-op speedup, for two reasons: (1) the per-tensor FP8 GEMM is only a part of the whole decode step, so it gets diluted by the rest of the computation; (2) at high concurrency the GPU approaches saturation and ops overlap, which further masks the GEMM's advantage.
